File: javascript/2024/solutions/day02part1.js
export const inputFile = '02.input';
export const testFile = '02.sample';
export const testOutputExpected = "2";

export function solve(input) {
    let safeReports = 0;

    input.split('\n').forEach(line => {
        let report = line.split(' ').map(x => parseInt(x));
        safeReports += reportIsSafe(report) ? 1 : 0;
    });
    
    return safeReports.toString();
}

function reportIsSafe(report) {
    let direction = 0;
    for (let i = 1; i < report.length; i++) {
        let diff = report[i] - report[i - 1];
        let isWithinRange = diff !== 0 && Math.abs(diff) <= 3;
        if (!isWithinRange) {
            return false;
        }
        if (direction === 0) {
            direction = diff > 0 ? 1 : -1;
            continue;
        }
        if (direction === 1 && diff < 0) {
            return false;
        }
        if (direction === -1 && diff > 0) {
            return false;
        }
    }
    return true;
}

File: csharp/2024/Solutions/Day02_Part1.cs
﻿namespace AdventOfCode.Solutions;

/// <summary>
/// https://adventofcode.com/2024/day/2
/// </summary>
internal sealed class Day02_Part1 : PuzzleSolution
{
    private const string DAY = "02";
    public static string FileName { get; } = $"Data/{DAY}.input";
    public static string TestFileName { get; } = $"Data/{DAY}.sample";
    public static string TestOutputExpected { get; } = "2";

    public static string Solve(StreamReader reader)
    {
        var safeReports = 0;
        while (reader.ReadLine() is { } line)
        {
            var numbers = line.Split(' ').Select(int.Parse).ToArray();
            var reportIsSafe = ReportIsSafe(numbers);
            if (reportIsSafe)
            {
                safeReports++;
            }
        }

        return safeReports.ToString();
    }

    private static bool ReportIsSafe(int[] numbers)
    {
        var direction = Change.None;
        for (var i = 1; i < numbers.Length; i++)
        {
            var diff = numbers[i] - numbers[i - 1];
            var isWithinRange = diff != 0 && Math.Abs(diff) <= 3;
            if (isWithinRange is false)
            {
                return false;
            }
            if (direction == Change.None)
            {
                direction = diff > 0 ? Change.Increasing : Change.Decreasing;
            }
            if (direction == Change.Increasing && diff < 0)
            {
                return false;
            }
            if (direction == Change.Decreasing && diff > 0)
            {
                return false;
            }
        }
        return true;
    }

    private enum Change
    {
        None = 0,
        Increasing = 1,
        Decreasing = -1
    }
}
